### PR TITLE
[SYCL] Fix ONEAPI_DEVICE_SELECTOR handling of discard filters.

### PR DESCRIPTION
**Pull request by uditagarwal97**
Currently, ONEAPI_DEVICE_SELECTOR incorrectly parses '!', used for discard filters.

### CODE REVIEW

File: sycl/test-e2e/OneapiDeviceSelector/discard_filters.cpp [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// REQUIRES: gpu
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple} %S/Inputs/trivial.cpp -o %t.out
+
+// Test discard filters in ONEAPI_DEVICE_SELECTOR.
+// RUN: env ONEAPI_DEVICE_SELECTOR="!*:gpu" %t.out | FileCheck %s --allow-empty
```
Comment (aelovikov-intel) [inline sycl/test-e2e/OneapiDeviceSelector/discard_filters.cpp:5]: Needs something like {run-unfiltered-devices}, don’t remember the exact spelling

File: sycl/source/detail/device_filter.cpp [hunk 2]
```diff
@@ -210,8 +212,24 @@ Parse_ONEAPI_DEVICE_SELECTOR(const std::string &envString) {
...
       throw sycl::exception(sycl::make_error_code(errc::invalid), ss.str());
-    } else if (Pair.size() == 2) {
-      backend be = Parse_ODS_Backend(Pair[0], Entry); // Pair[0] is backend.
+    } else if (Pair.size() > 2) {
+      std::stringstream ss;
+      ss << "Error parsing selector string \"" << Entry
+         << "\"  Too many colons (:)";
+      throw sycl::exception(sycl::make_error_code(errc::invalid), ss.str());
+    }
+
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/device_filter.cpp:221]: LGTM
NITs: 
1) why do we need extra empty line here?
2) last else if condition seems excessive - it is the only left case, we do not need a check
3) quering Pair.size() for every condition seems to be excessive too.

**Review by KseniyaTikhomirova — APPROVED** [review]

File: sycl/source/detail/device_filter.cpp (hunk 2 above)
Comment (uditagarwal97) [reply sycl/source/detail/device_filter.cpp:221]: 1. The empty line is to separate the erroneous `Pair` sizes from the acceptable one `Pair.size() == 2`.  Just to make the code a little cleaner.
2. True - I can remove the check from the last `else` condition.
3. I expect the optimizer to optimize multiple calls to `Pair.size()` as it is immutable after construction.

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers the PR is ready!

PRs on Friday evening: because who doesn’t love a good cliffhanger to start the weekend? :P

**Pull request merged**